### PR TITLE
Aseprite Import Support

### PR DESCRIPTION
**Pull request by aaaaaaaaargh**
This PR implements a basic import plugin that allows the workflow with aseprite files to be much more straightforward. You still need to set the aseprite binary via the config dialog, everything else can be done by using import settings on the given file directly.

I've implemented different modes to import the sprite data:

- SpriteFrames (default behaviour as used by the plugin) 
- Texture Strips (one single texture file per sprite/layer) 
- Texture Atlases (one AtlasTexture per animation along with textures for each frame)
- Animated textures (one AnimatedTexture per animation)

### CODE REVIEW

**Comment by viniciusgerevini** [comment]
Hello @aaaaaaaaargh. Thanks a lot for your contribution. I didn't know the image was saved as part of the resource file. That's quite clever. I actually had an importer as part of my initial project, but it was too buggy, so I removed it. This implementation looks neat and the extra texture options are really nice.

I noticed when changing the file it does not update the file in the editor, but I guess this is the same limitation I had with the wizard. Not much we can do and the game execution looks fine.

I was testing your changes and I have a few comments:

Can you leave the split option disabled by default? I believe having one Aseprite file as one resource is the most intuitive output for a starter.

The filename options for the extra texture files do not appear when selecting the checkbox, only after coming back to the importer dock. This seems to be a limitation in the importer's options visibility, which does not detect checkboxes changes. As a workaround, we could have the texture file option as a select field, with options like: None, Atlas Texture, Animated Texture and Texture strip, and then show the fields accordingly. What do you think about this?

I also noticed when re-importing with split off or without the texture selected does not remove the old files. I felt it a little bit weird, but I guess it's safer like this, right?

Thanks again for the great work.

**Comment by aaaaaaaaargh** [comment]
Hey @viniciusgerevini !

> Thanks a lot for your contribution. I didn't know the image was saved as part of the resource file. That's quite clever. I actually had an importer as part of my initial project, but it was too buggy, so I removed it. This implementation looks neat and the extra texture options are really nice.

You're very welcome. I've been using your plugin for a while now and I'm pretty much happy with it, so the least I can do is contribute my stuff back to it as long as it does make sense. That being said I've looking at pretty much all of the import plugins available for Aseprite and none seem to implement the EditorImportPlugin which -- at least for a originated Unity guy like me -- seems to be a little bit strange. I tried to keep this as modular as it can be, so I only made minimal intrusive changes to your codebase (hence the redundancy at some place I'm afraid). Anyways, this thing should be easy to incorporate, at least that's what was on my mind when doing it.

> I noticed when changing the file it does not update the file in the editor, but I guess this is the same limitation I had with the wizard. Not much we can do and the game execution looks fine.

Well, it does work _sometimes_ and I believe I found a way or another to circumvent this problem, but I didn't like these fixes too much as they felt very hacky to me and I don't like to throw something in there just for the sake of fixing a bug that is probably in the editor itself. I'd suggest to wait for 4.0 to be released and see how thing's are going then. Maybe I'm even ending getting deeper into Godot's development and contribute a fix there, but for now I think we should probably leave it as it is.


> 
> I was testing your changes and I have a few comments:
> 
> Can you leave the split option disabled by default? I believe having one Aseprite file as one resource is the most intuitive output for a starter.

Sure thing! That was probably a left-over from testing stuff. Consider it done, I already updated the PR.

> The filename options for the extra texture files do not appear when selecting the checkbox, only after coming back to the importer dock. This seems to be a limitation in the importer's options visibility, which does not detect checkboxes changes. As a workaround, we could have the texture file option as a select field, with options like: None, Atlas Texture, Animated Texture and Texture strip, and then show the fields accordingly. What do you think about this?

Yes, this is odd. I am not quite sure if I did something wrong here, because without the IDE updating option visibilities on changes I don't see the whole point of it at all. So the problem with a dropdown would be that you would only be able to select a single option whereas with checkboxes I can have multiple imports at once, which can be handy sometimes.

We could also just display all of the options at any time regardless of what was selected. This might not be the most elegant solution, but it surely wouldn't break anything.

Another idea was to use different presets, each one for a single import type. But I'm not an expert when it comes to these things, so I really don't know if it's that what presets are supposed to be. Again, as a Unity Guy my guess would be that this is more like a configuration template for different resource types, but I couldn't find any useful information on that. Also, even if it worked, we would end up with the same single-selection problem as with a select field and in that case I pretty much prefer your suggestion over this one.

> I also noticed when re-importing with split off or without the texture selected, the old files are not removed. I felt it a little bit weird, but I guess it's safer like this, right?

Yes, absolutely. You see, I tried to re-use your code as much as possible with as little change as possible. So what I ended up with was a create_resource call to aseprite via your command interface. That worked pretty well but it will leave me without any useful information about what layers could be extracted and so on. So what I do is simply check all of the files in the import folder and make the assumption that a .res file always corresponds with an exported layer. In pretty much all cases this seems to work just fine, but in order to delete files generated we might either set some kind of dependency correlation with something (probably the .aseprite main SpriteFrame resource) or do some book-keeping somewhere. For the latter, the only place I can think of that is not breaking when you move the file between projects or even folders would be the import metadata but I am not aware of any way one could save custom data in those files. So again, this left me pretty much without any viable alternative :/

> 
> Thanks again for the great work.

Well, thank you for starting this plugin and let us participate in what it can do :+1:

**Comment by viniciusgerevini** [comment]
> Yes, this is odd. I am not quite sure if I did something wrong here, because without the IDE updating option visibilities on changes I don't see the whole point of it at all. So the problem with a dropdown would be that you would only be able to select a single option whereas with checkboxes I can have multiple imports at once, which can be handy sometimes.

Yeah, I believe you did everything right. I had thought about the limitation of a dropdown and I was wondering how likely someone would want to have more than one form for same file at same time. Anyway, it is indeed a feature, and I don't mind leaving as is for the time being. We could maybe group and expose them, but this can be a improvement for later.  

> I tried to re-use your code as much as possible with as little change as possible

I noticed that and I'm glad you did it. The isolated changes makes it easier to review and I'm all about having a simpler solution working and then improving from that if required.

I'm happy with the current implementation. If you are Ok with that I can merge it and release it in the next version.

**Comment by aaaaaaaaargh** [comment]
> I'm happy with the current implementation. If you are Ok with that I can merge it and release it in the next version.

Yes of course, I'm glad you like it. Maybe I'll continue improving it from time to time, but as for now I think it's doing a pretty good job, too.

Thank you for reviewing, it's been fun discussing these things.

**Comment by viniciusgerevini** [comment]
Sure thing! Thanks o/

**Comment by viniciusgerevini** [comment]
Changes already landed on version 1.1.0 in the Asset Library. Thanks again for the contribution. 🥳

**Comment by aaaaaaaaargh** [comment]
You're very welcome. Thanks again for starting this great project!

By the way, I tried to improve the caching and in-editor updates today. I had quite some success, but the way of how things need to be implemented are extremely confusing because you need to deal with .stex files and invalidation of existing resources. I may be able to improve the current state a little bit, but to be honest everything feels like its making things even more worse, so we probably still want for 4.0 to be released an see what happens then.

**Comment by viniciusgerevini** [comment]
Nice. Thanks for that. Yeah, I also tried to look into the caching issue, and things get quite hacky very quickly lol. I think as long as it doesn't impact running the game, it should be fine to have it as it is for now.

**Comment by aaaaaaaaargh** [comment]
Agreed. And no, I think the game is completely unaffected by this because it doesn't use any of the editors cached resources (at least i hope so :smile:  ). The problem seems to be that you need to update the resource as residing in cache instead of saving a new one. That, however, leads to a whole new problem, because as I learned Godot seems to save ImageTexture resources a StreamingTextures internally. You can verify that by looking at the contents of the .import folder where you should see an awful lot of .stex files. As a result we should create StreamTextures instead of images but that requires us to write the .stex files manually. It's really not that complicated because its a very simple (yet binary) format and I had successfully been doing that in a testing branch. However, the big show stopper seems to be the SpriteFrames instance. I have not been able to update this one regardless of what I tried.

So, long story short, as long as we need to go down this road I suggest not to do it. It's doable, sure, but if you ask me it's just way too hacky and blows up the complexity of the implementeation in an unjustifiable manner.

**Comment by viniciusgerevini** [comment]
Fair enough. I totally agree. Thanks for spending the time on that :)
